File: ios/SsdidWallet/Domain/OpenId4Vci/OpenId4VciHandler.swift
import Foundation

/// Error types for OpenID4VCI operations.
enum OpenId4VciError: Error, LocalizedError {
    case invalidOffer(String)
    case metadataError(String)
    case tokenError(String)
    case transportError(String)
    case proofError(String)
    case credentialError(String)
    case unsupported(String)

    var errorDescription: String? {
        switch self {
        case .invalidOffer(let reason):
            return "Invalid credential offer: \(reason)"
        case .metadataError(let reason):
            return "Metadata error: \(reason)"
        case .tokenError(let reason):
            return "Token error: \(reason)"
        case .transportError(let reason):
            return "VCI transport error: \(reason)"
        case .proofError(let reason):
            return "Proof error: \(reason)"
        case .credentialError(let reason):
            return "Credential error: \(reason)"
        case .unsupported(let reason):
            return "Unsupported: \(reason)"
        }
    }
}

/// Result of processing a credential offer before user consent.
struct CredentialOfferReview {
    let offer: CredentialOffer
    let metadata: IssuerMetadata
    let credentialConfigNames: [String]
}

/// Result of a credential issuance attempt.
enum IssuanceResult {
    case success(StoredSdJwtVc)
    case mdocSuccess(StoredMDoc)
    case deferred(transactionId: String, deferredEndpoint: String, accessToken: String)
    case failed(String)
}

/// Protocol for storing SD-JWT VCs, to be implemented by vault storage.
protocol SdJwtVcStorage {
    func saveSdJwtVc(_ sdJwtVc: StoredSdJwtVc) async throws
}

/// Protocol for storing mdoc credentials, to be implemented by vault storage.
protocol MDocStorage {
    func saveMDoc(_ mdoc: StoredMDoc) async throws
}

/// Orchestrates the OpenID4VCI credential issuance flow.
final class OpenId4VciHandler {

    private let metadataResolver: IssuerMetadataResolver
    private let tokenClient: TokenClient
    private let nonceManager: NonceManager
    private let transport: OpenId4VciTransport
    private let vcStorage: SdJwtVcStorage
    private let mdocStorage: MDocStorage?

    init(
        metadataResolver: IssuerMetadataResolver,
        tokenClient: TokenClient,
        nonceManager: NonceManager,
        transport: OpenId4VciTransport,
        vcStorage: SdJwtVcStorage,
        mdocStorage: MDocStorage? = nil
    ) {
        self.metadataResolver = metadataResolver
        self.tokenClient = tokenClient
        self.nonceManager = nonceManager
        self.transport = transport
        self.vcStorage = vcStorage
        self.mdocStorage = mdocStorage
    }

    /// Processes a credential offer URI: parses, resolves metadata, and returns review data.
    func processOffer(uri: String) async throws -> CredentialOfferReview {
        let offer = try CredentialOffer.parseFromUri(uri)
        let metadata = try await metadataResolver.resolve(issuerUrl: offer.credentialIssuer)

        return CredentialOfferReview(
            offer: offer,
            metadata: metadata,
            credentialConfigNames: offer.credentialConfigurationIds
        )
    }

    /// Accepts a credential offer and performs the issuance flow.
    func acceptOffer(
        offer: CredentialOffer,
        metadata: IssuerMetadata,
        selectedConfigId: String,
        txCode: String? = nil,
        walletDid: String,
        keyId: String,
        algorithm: String,
        signer: @escaping (Data) -> Data
    ) async throws -> IssuanceResult {
        // Step 1: Token exchange
        guard let preAuthCode = offer.preAuthorizedCode else {
            throw OpenId4VciError.unsupported(
                "Authorization code flow requires browser interaction"
            )
        }

        let tokenResponse = try await tokenClient.exchangePreAuthorizedCode(
            tokenEndpoint: metadata.tokenEndpoint,
            preAuthorizedCode: preAuthCode,
            txCode: txCode
        )

        // Step 2: Update nonce from token response
        if let nonce = tokenResponse.cNonce {
            await nonceManager.update(
                nonce: nonce,
                expiresIn: tokenResponse.cNonceExpiresIn ?? 300
            )
        }

        // Step 3: Request credential
        return try await requestCredential(
            metadata: metadata,
            accessToken: tokenResponse.accessToken,
            selectedConfigId: selectedConfigId,
            walletDid: walletDid,
            keyId: keyId,
            algorithm: algorithm,
            signer: signer
        )
    }

    // MARK: - Private

    private func requestCredential(
        metadata: IssuerMetadata,
        accessToken: String,
        selectedConfigId: String,
        walletDid: String,
        keyId: String,
        algorithm: String,
        signer: (Data) -> Data
    ) async throws -> IssuanceResult {
        guard let currentNonce = await nonceManager.current() else {
            throw OpenId4VciError.credentialError("No c_nonce available")
        }

        let proofJwt = try ProofJwtBuilder.build(
            algorithm: algorithm,
            keyId: keyId,
            walletDid: walletDid,
            issuerUrl: metadata.credentialIssuer,
            nonce: currentNonce,
            signer: signer
        )

        // Determine format from credential configuration
        let configObj = metadata.credentialConfigurationsSupported[selectedConfigId]
        let format = configObj?["format"] as? String ?? "vc+sd-jwt"
        let isMDoc = format == "mso_mdoc"

        let requestBody: [String: Any]
        if isMDoc {
            let doctype = configObj?["doctype"] as? String ?? selectedConfigId
            requestBody = [
                "format": "mso_mdoc",
                "doctype": doctype,
                "proof": [
                    "proof_type": "jwt",
                    "jwt": proofJwt
                ]
            ]
        } else {
            requestBody = [
                "format": "vc+sd-jwt",
                "credential_definition": ["vct": selectedConfigId],
                "proof": [
                    "proof_type": "jwt",
                    "jwt": proofJwt
                ]
            ]
        }

        let requestBodyData = try JSONSerialization.data(withJSONObject: requestBody)
        guard let requestBodyString = String(data: requestBodyData, encoding: .utf8) else {
            throw OpenId4VciError.credentialError("Failed to serialize request body")
        }

        let responseStr = try await transport.postCredentialRequest(
            credentialEndpoint: metadata.credentialEndpoint,
            accessToken: accessToken,
            requestBody: requestBodyString
        )

        guard let responseData = responseStr.data(using: .utf8),
              let response = try JSONSerialization.jsonObject(with: responseData) as? [String: Any] else {
            throw OpenId4VciError.credentialError("Invalid credential response JSON")
        }

        // Check for credential in response
        if let credential = response["credential"] as? String {
            // Update nonce from response if present
            if let nonce = response["c_nonce"] as? String {
                let expiresIn = response["c_nonce_expires_in"] as? Int ?? 300
                await nonceManager.update(nonce: nonce, expiresIn: expiresIn)
            }

            if isMDoc {
                return try await handleMDocCredential(
                    credentialBase64: credential,
                    selectedConfigId: selectedConfigId,
                    keyId: keyId,
                    metadata: metadata
                )
            }

            let storedVc = StoredSdJwtVc(
                id: UUID().uuidString,
                compact: credential,
                issuer: metadata.credentialIssuer,
                subject: walletDid,
                type: selectedConfigId,
                claims: [:],
                disclosableClaims: [],
                issuedAt: Int64(Date().timeIntervalSince1970)
            )

            try await vcStorage.saveSdJwtVc(storedVc)
            return .success(storedVc)
        }

        // Check for deferred issuance
        if let transactionId = response["transaction_id"] as? String {
            let deferredEndpoint = metadata.credentialEndpoint
                .replacingOccurrences(of: "/credential", with: "/deferred_credential")
            return .deferred(
                transactionId: transactionId,
                deferredEndpoint: deferredEndpoint,
                accessToken: accessToken
            )
        }

        return .failed("Unexpected response from credential endpoint")
    }

    private func handleMDocCredential(
        credentialBase64: String,
        selectedConfigId: String,
        keyId: String,
        metadata: IssuerMetadata
    ) async throws -> IssuanceResult {
        guard let cborBytes = Data(base64Encoded: credentialBase64
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        ) else {
            throw OpenId4VciError.credentialError("Invalid base64url credential data")
        }

        let decoded = CborCodec.decodeMap(cborBytes) ?? [:]
        let docType = decoded["docType"] as? String ?? selectedConfigId

        // Extract namespace keys from decoded CBOR
        let nameSpacesRaw = decoded["nameSpaces"] as? [String: Any] ?? [:]
        var nameSpaces = [String: [String]]()
        for (ns, value) in nameSpacesRaw {
            if let arr = value as? [Any] {
                nameSpaces[ns] = arr.compactMap { $0 as? String }
            } else if let dict = value as? [String: Any] {
                nameSpaces[ns] = Array(dict.keys)
            }
        }

        let configDoctype = metadata.credentialConfigurationsSupported[selectedConfigId]?["doctype"] as? String

        let storedMDoc = StoredMDoc(
            id: UUID().uuidString,
            docType: configDoctype ?? docType,
            issuerSignedCbor: cborBytes,
            deviceKeyId: keyId,
            issuedAt: Int64(Date().timeIntervalSince1970),
            nameSpaces: nameSpaces
        )

        if let mdocStorage = mdocStorage {
            try await mdocStorage.saveMDoc(storedMDoc)
        }
        return .mdocSuccess(storedMDoc)
    }
}

File: android/app/src/main/java/my/ssdid/wallet/domain/oid4vci/OpenId4VciHandler.kt
package my.ssdid.wallet.domain.oid4vci

import kotlinx.serialization.json.*
import my.ssdid.wallet.domain.mdoc.CborCodec
import my.ssdid.wallet.domain.mdoc.StoredMDoc
import my.ssdid.wallet.domain.sdjwt.StoredSdJwtVc
import my.ssdid.wallet.domain.vault.Vault

data class CredentialOfferReview(
    val offer: CredentialOffer,
    val metadata: IssuerMetadata,
    val credentialConfigNames: List<String>
)

sealed class IssuanceResult {
    data class Success(val credential: StoredSdJwtVc) : IssuanceResult()
    data class MDocSuccess(val mdoc: StoredMDoc) : IssuanceResult()
    class Deferred(
        val transactionId: String,
        val deferredEndpoint: String,
        val accessToken: String
    ) : IssuanceResult() {
        override fun toString() = "Deferred(transactionId=$transactionId, deferredEndpoint=$deferredEndpoint, accessToken=<redacted>)"
    }
    data class Failed(val error: String) : IssuanceResult()
}

class OpenId4VciHandler(
    private val metadataResolver: IssuerMetadataResolver,
    private val tokenClient: TokenClient,
    private val nonceManager: NonceManager,
    private val transport: OpenId4VciTransport,
    private val vault: Vault
) {
    private val json = Json { ignoreUnknownKeys = true }

    fun processOffer(uri: String): Result<CredentialOfferReview> = runCatching {
        val offer = CredentialOffer.parseFromUri(uri).getOrThrow()
        val metadata = metadataResolver.resolve(offer.credentialIssuer).getOrThrow()

        CredentialOfferReview(
            offer = offer,
            metadata = metadata,
            credentialConfigNames = offer.credentialConfigurationIds
        )
    }

    suspend fun acceptOffer(
        offer: CredentialOffer,
        metadata: IssuerMetadata,
        selectedConfigId: String,
        txCode: String? = null,
        walletDid: String,
        keyId: String,
        algorithm: String,
        signer: (ByteArray) -> ByteArray
    ): Result<IssuanceResult> = runCatching {
        // Step 1: Token exchange
        val tokenResponse = if (offer.preAuthorizedCode != null) {
            tokenClient.exchangePreAuthorizedCode(
                metadata.tokenEndpoint,
                offer.preAuthorizedCode,
                txCode
            ).getOrThrow()
        } else {
            throw UnsupportedOperationException(
                "Authorization code flow requires browser interaction"
            )
        }

        // Step 2: Update nonce from token response
        tokenResponse.cNonce?.let { nonce ->
            nonceManager.update(nonce, tokenResponse.cNonceExpiresIn ?: 300)
        }

        // Step 3: Request credential
        requestCredential(
            metadata = metadata,
            accessToken = tokenResponse.accessToken,
            selectedConfigId = selectedConfigId,
            walletDid = walletDid,
            keyId = keyId,
            algorithm = algorithm,
            signer = signer
        )
    }

    private suspend fun requestCredential(
        metadata: IssuerMetadata,
        accessToken: String,
        selectedConfigId: String,
        walletDid: String,
        keyId: String,
        algorithm: String,
        signer: (ByteArray) -> ByteArray
    ): IssuanceResult {
        val currentNonce = nonceManager.current()
            ?: throw IllegalStateException("No c_nonce available")

        val proofJwt = ProofJwtBuilder.build(
            algorithm = algorithm,
            keyId = keyId,
            walletDid = walletDid,
            issuerUrl = metadata.credentialIssuer,
            nonce = currentNonce,
            signer = signer
        )

        // Determine format from credential configuration
        val configObj = metadata.credentialConfigurationsSupported[selectedConfigId]
        val format = configObj?.get("format")?.jsonPrimitive?.contentOrNull ?: "vc+sd-jwt"
        val isMDoc = format == "mso_mdoc"

        val requestBody = if (isMDoc) {
            val doctype = configObj?.get("doctype")?.jsonPrimitive?.contentOrNull
                ?: selectedConfigId
            buildJsonObject {
                put("format", "mso_mdoc")
                put("doctype", doctype)
                putJsonObject("proof") {
                    put("proof_type", "jwt")
                    put("jwt", proofJwt)
                }
            }
        } else {
            buildJsonObject {
                put("format", "vc+sd-jwt")
                putJsonObject("credential_definition") {
                    put("vct", selectedConfigId)
                }
                putJsonObject("proof") {
                    put("proof_type", "jwt")
                    put("jwt", proofJwt)
                }
            }
        }

        val responseStr = transport.postCredentialRequest(
            metadata.credentialEndpoint,
            accessToken,
            requestBody.toString()
        )

        val response = json.parseToJsonElement(responseStr).jsonObject

        // Check for credential in response
        val credential = response["credential"]?.jsonPrimitive?.contentOrNull
        if (credential != null) {
            // Update nonce from response if present
            response["c_nonce"]?.jsonPrimitive?.contentOrNull?.let { nonce ->
                nonceManager.update(
                    nonce,
                    response["c_nonce_expires_in"]?.jsonPrimitive?.intOrNull ?: 300
                )
            }

            if (isMDoc) {
                return handleMDocCredential(credential, selectedConfigId, keyId, metadata)
            }

            val storedVc = StoredSdJwtVc(
                id = java.util.UUID.randomUUID().toString(),
                compact = credential,
                issuer = metadata.credentialIssuer,
                subject = walletDid,
                type = selectedConfigId,
                claims = emptyMap(),
                disclosableClaims = emptyList(),
                issuedAt = System.currentTimeMillis() / 1000
            )

            vault.storeStoredSdJwtVc(storedVc)
            return IssuanceResult.Success(storedVc)
        }

        // Check for deferred issuance
        val transactionId = response["transaction_id"]?.jsonPrimitive?.contentOrNull
        if (transactionId != null) {
            return IssuanceResult.Deferred(
                transactionId = transactionId,
                deferredEndpoint = metadata.credentialEndpoint
                    .replace("/credential", "/deferred_credential"),
                accessToken = accessToken
            )
        }

        return IssuanceResult.Failed("Unexpected response from credential endpoint")
    }

    private suspend fun handleMDocCredential(
        credentialBase64: String,
        selectedConfigId: String,
        keyId: String,
        metadata: IssuerMetadata
    ): IssuanceResult.MDocSuccess {
        val cborBytes = java.util.Base64.getUrlDecoder().decode(credentialBase64)

        // Parse the IssuerSigned to extract element identifiers per namespace
        val issuerSigned = my.ssdid.wallet.domain.mdoc.MsoParser.parseIssuerSigned(cborBytes)
        val decoded = CborCodec.decodeMap(cborBytes)

        @Suppress("UNCHECKED_CAST")
        val docType = decoded["docType"] as? String ?: selectedConfigId

        // Extract element identifiers from parsed IssuerSignedItems
        val nameSpaces = issuerSigned.nameSpaces.mapValues { (_, items) ->
            items.map { it.elementIdentifier }
        }

        val configObj = metadata.credentialConfigurationsSupported[selectedConfigId]
        val configDoctype = configObj?.get("doctype")?.jsonPrimitive?.contentOrNull

        val storedMDoc = StoredMDoc(
            id = java.util.UUID.randomUUID().toString(),
            docType = configDoctype ?: docType,
            issuerSignedCbor = cborBytes,
            deviceKeyId = keyId,
            issuedAt = System.currentTimeMillis() / 1000,
            nameSpaces = nameSpaces
        )

        vault.storeMDoc(storedMDoc)
        return IssuanceResult.MDocSuccess(storedMDoc)
    }
}
